src/exodus/sigmawallet.cpp

```diff
@@ -0,0 +1,597 @@
+    auto serialId = GetSerialId(coin.serial);
+    mint = SigmaMint(
+        propertyId,
+        denomination,
+        mintPoolEntry->seedId,
+        serialId
+    );
+
+    LogPrintf("%s: pubcoin: %s\n", __func__, key.commitment.GetHex());
+    return true;
+}
+
+SigmaMint SigmaWallet::UpdateMint(const SigmaMintId &id, const std::function<void(SigmaMint &)> &modifier)
+{
+    CWalletDB walletdb(walletFile);
+    auto m = GetMint(id);
+    modifier(m);
+
+    if (!walletdb.WriteExodusHDMint(id, m)) {
+        throw std::runtime_error("fail to update mint");
+    }
+
+    return m;
+}
+
+bool SigmaWallet::RegenerateMint(const SigmaMint& mint, SigmaPrivateKey &privKey)
+{
+    SigmaMint dummyMint;
+
+    auto priv = GetPrivateKeyFromSeedId(mint.seedId);
+    SigmaPublicKey pub(priv, DefaultSigmaParams);
+
+    MintPoolEntry mintPoolEntry(pub, mint.seedId);
+    if (!GenerateMint(mint.property, mint.denomination, privKey, dummyMint, mintPoolEntry)) {
+        return error("%s: failed to generate mint", __func__);
+    }
+
+    // Verify regenered
+    exodus::SigmaPublicKey pubKey(privKey, DefaultSigmaParams);
+    if (pubKey != pub) {
+        return error("%s: failed to correctly generate mint, pubcoin mismatch", __func__);
+    }
+
+    if (GetSerialId(privKey.serial) != mint.serialId) {
+        return error("%s: failed to correctly generate mint, serial hash mismatch", __func__);
+    }
+
+    return true;
+}
+
+void SigmaWallet::ResetCoinsState()
+{
+    try {
+        CWalletDB walletdb(walletFile);
+
+        ListSigmaMints([&](SigmaMint &m) {
+
+            m.chainState = SigmaMintChainState();
+            m.spendTx = uint256();
+
+            auto priv = GetPrivateKeyFromSeedId(m.seedId);
+            SigmaPublicKey pub(priv, DefaultSigmaParams);
+
+            if (!walletdb.WriteExodusHDMint(
+                SigmaMintId(m.property, m.denomination, pub), m)) {
+
+               throw std::runtime_error("fail to update hdmint");
+            }
+
+        }, false, false);
+
+    } catch (std::runtime_error const &e) {
+        LogPrintf("%s : fail to reset all mints chain state, %s\n", __func__, e.what());
+        throw;
+    }
+}
+
+bool SigmaWallet::SetMintSeedSeen(
+    MintPoolEntry const &mintPoolEntry,
+    uint32_t propertyId,
+    uint8_t denomination,
+    exodus::SigmaMintChainState const &chainState,
+    uint256 const &spendTx)
+{
+    // Regenerate the mint
+    auto const &pubcoin = mintPoolEntry.key;
+    auto const &seedId = mintPoolEntry.seedId;
+    auto seedIndex = GetSeedIndex(seedId);
+
+    SigmaMintId id(propertyId, denomination, pubcoin);
+
+    uint160 serialId;
+
+    // Can regenerate if unlocked (cheaper)
+    if (!pwalletMain->IsLocked()) {
+
+        uint512 seed;
+        GenerateSeed(seedId, seed);
+
+        SigmaPrivateKey coin;
+        if (!SeedToPrivateKey(seed, coin)) {
+            return false;
+        }
+
+        serialId = GetSerialId(coin.serial);
+    } else {
```

Comment:
The only cases that will invoke this method is when the mint not created yet but it already in the pool. That mean we don't need this case.